pkg/controller/statefulset/stateful_set_control.go

```diff
@@ -393,94 +395,90 @@ func (ssc *defaultStatefulSetControl) updateStatefulSet(
-				"StatefulSet %s/%s is recreating failed Pod %s",
-				set.Namespace,
-				set.Name,
-				replicas[i].Name)
-			if err := ssc.podControl.DeleteStatefulPod(set, replicas[i]); err != nil {
-				return &status, err
-			}
-			if getPodRevision(replicas[i]) == currentRevision.Name {
-				status.CurrentReplicas--
-			}
-			if getPodRevision(replicas[i]) == updateRevision.Name {
-				status.UpdatedReplicas--
-			}
-			status.Replicas--
-			replicaOrd := i + getStartOrdinal(set)
-			replicas[i] = newVersionedStatefulSetPod(
-				currentSet,
-				updateSet,
-				currentRevision.Name,
-				updateRevision.Name,
-				replicaOrd)
-		}
-		// If we find a Pod that has not been created we create the Pod
-		if !isCreated(replicas[i]) {
-			if utilfeature.DefaultFeatureGate.Enabled(features.StatefulSetAutoDeletePVC) {
-				if isStale, err := ssc.podControl.PodClaimIsStale(set, replicas[i]); err != nil {
-					return &status, err
-				} else if isStale {
-					// If a pod has a stale PVC, no more work can be done this round.
-					return &status, err
+		// Make i per iteration (not per loop) variable.
+		i := i
+		if isFailed(replicas[i]) || !isCreated(replicas[i]) {
+			fns = append(fns, func() (bool, error) {
+				// delete and recreate failed pods
+				if isFailed(replicas[i]) {
+					ssc.recorder.Eventf(set, v1.EventTypeWarning, "RecreatingFailedPod",
+						"StatefulSet %s/%s is recreating failed Pod %s",
+						set.Namespace,
+						set.Name,
+						replicas[i].Name)
+					if err := ssc.podControl.DeleteStatefulPod(set, replicas[i]); err != nil {
+						return true, err
+					}
+					replicaOrd := i + getStartOrdinal(set)
+					replicas[i] = newVersionedStatefulSetPod(
+						currentSet,
+						updateSet,
+						currentRevision.Name,
+						updateRevision.Name,
+						replicaOrd)
 				}
-			}
-			if err := ssc.podControl.CreateStatefulPod(ctx, set, replicas[i]); err != nil {
-				return &status, err
-			}
-			status.Replicas++
-			if getPodRevision(replicas[i]) == currentRevision.Name {
-				status.CurrentReplicas++
-			}
-			if getPodRevision(replicas[i]) == updateRevision.Name {
-				status.UpdatedReplicas++
-			}
-			// if the set does not allow bursting, return immediately
+				// If we find a Pod that has not been created we create the Pod
+				if !isCreated(replicas[i]) {
+					if utilfeature.DefaultFeatureGate.Enabled(features.StatefulSetAutoDeletePVC) {
+						if isStale, err := ssc.podControl.PodClaimIsStale(set, replicas[i]); err != nil {
+							return true, err
+						} else if isStale {
+							// If a pod has a stale PVC, no more work can be done this round.
+							return true, err
+						}
+					}
+					return true, ssc.podControl.CreateStatefulPod(ctx, set, replicas[i])
+				}
+				return true, nil
+			})
+
 			if monotonic {
-				return &status, nil
+				// if the set does not allow bursting, end loop - no point stacking more functions as they won't be executed anyway.
```

Comment:
Isn't this break sufficient to get rid of the "shouldExit" return value?

IIUC we return "true, X" only in two places:
* here
* delete statefulset pod part

In both cases you break right after that so there are no functions in fns added after fn with potential shouldExit = true value.

"if monotonic == true", you have to exist function after executing fns, but before you go to the next loop (e.g. pod deletion) so you can focus on a single for loop.

Reply:
It's not after the fix to above comments about exiting outer function in monotonic mode.

I also think it's easier to follow the code when this value is returned explicitly.